master/internal/stream/testutils.go

```diff
@@ -0,0 +1,234 @@
+	return i
+}
+
+// mockSocket implements WebsocketLike and stores all messages received from streaming.
+type mockSocket struct {
+	inbound  chan interface{}
+	outbound chan *StartupMsg
+	closed   chan struct{}
+}
+
+// newMockSocket creates a new instance mockSocket and initialize it's conditional variables.
+func newMockSocket() *mockSocket {
+	return &mockSocket{
+		inbound:  make(chan interface{}, channelBufferSize),
+		outbound: make(chan *StartupMsg, channelBufferSize),
+		closed:   make(chan struct{}),
+	}
+}
+
+// WriteOut synchrounously appends the StartupMsg to outbound messages.
+func (s *mockSocket) WriteOutbound(startup *StartupMsg) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case s.outbound <- startup:
+		return nil
+	}
+}
+
+// ReadJSON implements WebsocketLike's ReadJSON(), blocks until able to read outbound messages off the mockSocket
+// or the mockSocket is closed.
+func (s *mockSocket) ReadJSON(data interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case msg := <-s.outbound:
+		targetMsg, ok := data.(*StartupMsg)
+		if !ok {
+			return fmt.Errorf("target message type is not a pointer to StartupMsg")
+		}
+		*targetMsg = *msg
+		return nil
+	}
+}
+
+// Write implements WebsocketLike's Write(), appends the data to mockSocket's inbound messages.
+func (s *mockSocket) Write(data interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case s.inbound <- data:
+		return nil
+	}
+}
+
+// ReadData synchrounously reads an inbound message off the mockSocket.
+func (s *mockSocket) ReadIncoming(data *interface{}) error {
+	select {
+	case <-s.closed:
+		return &websocket.CloseError{Code: websocket.CloseAbnormalClosure}
+	case msg := <-s.inbound:
+		*data = msg
+		return nil
+	}
+}
+
+// ReadUntil reads until the terminationMsg has been read.
+func (s *mockSocket) ReadUntil(data *[]interface{}, terminationMsg interface{}) error {
+	var msg interface{}
+	for {
+		if reflect.TypeOf(msg) == reflect.TypeOf(terminationMsg) {
+			switch typedMsg := msg.(type) {
+			case stream.UpsertMsg:
+				if err := validateUpsertMsg(typedMsg.Msg, terminationMsg.(stream.UpsertMsg).Msg); err == nil {
+					return nil
+				}
+			case string:
+				if typedMsg == terminationMsg.(string) {
+					return nil
+				}
+			case SyncMsg:
+				if typedMsg.SyncID == terminationMsg.(SyncMsg).SyncID {
+					return nil
+				}
+			}
+		}
+		if err := s.ReadIncoming(&msg); err != nil {
+			return err
+		}
+		*data = append(*data, msg)
+	}
+}
+
+// Close closes the mockSocket.
+func (s *mockSocket) Close() {
+	close(s.closed)
+}
+
+func splitMsgs[M stream.Msg](messages []interface{}) (
+	deletions []string,
+	upserts []stream.Msg,
+	syncs []SyncMsg,
+	err error,
+) {
+	typeHolder := new(M)
+	for _, msg := range messages {
+		if deletion, ok := msg.(stream.DeleteMsg); ok {
```

Comment:
Good call out! This will make it more consistent with the other helper functions. Plus also after watching the advanced testing video, it makes more sense to have `t` passed in, and we have the test fail, rather than return an error